src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp

```diff
@@ -17,39 +19,75 @@ namespace RelativisticEuler {
 namespace Valencia {
 
 template <size_t Dim>
-std::array<DataVector, Dim + 2> characteristic_speeds(
+void characteristic_speeds(
+    const gsl::not_null<std::array<DataVector, Dim + 2>*> char_speeds,
     const Scalar<DataVector>& lapse, const tnsr::I<DataVector, Dim>& shift,
     const tnsr::I<DataVector, Dim>& spatial_velocity,
     const Scalar<DataVector>& spatial_velocity_squared,
     const Scalar<DataVector>& sound_speed_squared,
     const tnsr::i<DataVector, Dim>& normal) noexcept {
-  auto characteristic_speeds =
-      make_array<Dim + 2>(DataVector(-1.0 * get(dot_product(normal, shift))));
+  const size_t num_grid_points = get<0>(shift).size();
+  constexpr size_t num_vectors = 5;
+  std::unique_ptr<double[], decltype(&free)> temp_vectors(
```

Comment:
In future, if this is done frequently, would it be not be preferable to have a buffer-management class similar to a thin version of `Variables`? I think it would be a less concerning policy as a whole to encourage development using buffers obtained from a class that has been tested to have safe memory management, rather than to have optimized code all separately implement manual memory management.

Reply:
`unique_ptr` does memory management.

Reply:
of course it does :), but do you really want to encourage us to develop physics computations with mallocs and pointer offsets? There's good reason clang-tidy complains...

Reply:
Nils: your first reply in this thread looks like it's missing the end of the first sentence. I can't tell whether you liked or disliked adding a code comment like the one at the end of the reply. 

Personally, I would find your suggested brief comment helpful. I agree the long explanation can't (and shouldn't) be given every time the code is optimized. But a short comment (exactly as you've written above) saying "we do this non-intuitive thing for optimization" makes it clearer what the thought process is.

Reply:
Yes, I'm in favor of add the brief.

@moxcodes there are plans for a std::array+Variables style data structures.

Reply:
The reason clang-tidy complains is incorrect. It doesn't want you to do:
```cpp
double* blah = malloc(...);
```
because of potential memory leaks. `unique_ptr` remedies this, so in this case clang-tidy is giving a false positive. However, I agree about the offsets :)